third_party/nvfuser/runtime/helpers.cu

```diff
@@ -420,6 +420,67 @@ __device__ int64_t pow(int a, int64_t b) {
   return pow((int64_t)a, b);
 }
 
+// The reciprocal of a complex number z is
+//    1/z = conj(z)/|z|^2.
+// The principal square root of a complex number z can be obtained by [1]
+//    sqrt(z) = sqrt(|z|) (z + |z|) / |z + |z||.
+// Combining these formulas we have
+//    1/sqrt(z) = (conj(z) + |z|) / (sqrt(|z|) |z + |z||).
+// [1] https://math.stackexchange.com/a/44500
+__device__ std::complex<float> rsqrt(std::complex<float> z) {
+  auto a = std::real(z);
+  auto b = std::imag(z);
+  auto absa = ::fabsf(a);
+  auto absb = ::fabsf(b);
+  // scale to avoid precision loss due to underflow/overflow
+  auto scale = fmax(absa, absb);
+  a /= scale;
+  b /= scale;
+  auto a_sq = a * a;
+  auto b_sq = b * b;
+  auto modz_sq = a_sq + b_sq;
+  auto modz = ::sqrtf(modz_sq);
+  auto a_plus_modz = a + modz;
+  auto mod_zplusmodz_sq = a_plus_modz * a_plus_modz + b_sq;
+  auto fac = ::rsqrtf(scale * modz * mod_zplusmodz_sq);
+  return std::complex<float>(a_plus_modz * fac, -b * fac);
+}
+
+__device__ std::complex<double> rsqrt(std::complex<double> z) {
+  auto a = std::real(z);
+  auto b = std::imag(z);
+  auto absa = ::abs(a);
+  auto absb = ::abs(b);
+  // scale to avoid precision loss due to underflow/overflow
+  auto scale = fmax(absa, absb);
+  a /= scale;
+  b /= scale;
+  auto a_sq = a * a;
+  auto b_sq = b * b;
+  auto modz_sq = a_sq + b_sq;
+  auto modz = ::sqrt(modz_sq);
+  auto a_plus_modz = a + modz;
+  auto mod_zplusmodz_sq = a_plus_modz * a_plus_modz + b_sq;
+  auto fac = ::rsqrt(scale * modz * mod_zplusmodz_sq);
+  return std::complex<double>(a_plus_modz * fac, -b * fac);
+}
+
+__device__ double rsqrt(double z) {
```

Comment:
Out of curiosity, do we need these for name lookup? This is subtle...

Reply:
Yeah since this is named rsqrt in present scope it shadows the nvcc builtin and nvcc was complaining. The `::` silenced it.